### ISSUE TITLE
Move commit message search out of critical chain

### ISSUE THREAD

**Issue by usrme** [usrme/comet-alt#10]
While the functionality added in #9 has already proved useful to me, I have noticed that in some scenarios the start-up of Cometary does wait for just a little bit. Not every time, but often enough to irk me. This is because the [messages are searched](https://github.com/usrme/cometary/blob/dec06c48d1c964c517767fbf393b2f4439064339/main.go#L44-L50) for prior to drawing any user interface.

If the searching was deferred to a less critical portion of the program's lifecycle, then the user experience (i.e. my experience...) would be improved. As far as I'm concerned the loading should just be done by the time the actual commit message is being input and if it's still searching, then a message should be shown.

The following are some rough benchmarks from [Hyperfine](https://github.com/sharkdp/hyperfine). I'm using `--output inherit` because it's [not possible to redirect output from an interactive Bubbles program](https://github.com/charmbracelet/bubbletea/discussions/663) and `--ignore-failure` because I'm running the benchmarks on a custom build that quits the program immediately after drawing the interface.

---

Private repository with ~2000 commits:

```console
$ hyperfine --warmup 10 --output inherit --ignore-failure 'cometary' 'cometary -m fix'
Time (mean ± σ):      29.8 ms ±   3.8 ms    [User: 7.8 ms, System: 15.2 ms]
Range (min … max):    24.8 ms …  43.5 ms    106 runs
...
Time (mean ± σ):      72.4 ms ±  16.1 ms    [User: 37.9 ms, System: 37.6 ms]
Range (min … max):    51.2 ms … 102.6 ms    49 runs

Summary
  cometary ran
    2.51 ± 0.58 times faster than cometary -m fix
```

['flyctl'](https://github.com/superfly/flyctl) repository with ~6000 commits:

```console
$ hyperfine --warmup 10 --output inherit --ignore-failure 'cometary' 'cometary -m fix'
Time (mean ± σ):      29.9 ms ±   4.4 ms    [User: 8.1 ms, System: 14.7 ms]
Range (min … max):    24.3 ms …  48.6 ms    91 runs
...
Time (mean ± σ):     111.4 ms ±  15.9 ms    [User: 87.4 ms, System: 36.6 ms]
Range (min … max):    83.2 ms … 133.7 ms    26 runs

Summary
  cometary ran
    3.72 ± 0.77 times faster than cometary -m fix
```

['kubernetes'](https://github.com/kubernetes/kubernetes) repository with ~117K commits:

```console
$ hyperfine --warmup 10 --output inherit --ignore-failure 'cometary' 'cometary -m fix'
Time (mean ± σ):      39.7 ms ±   6.8 ms    [User: 14.9 ms, System: 23.7 ms]
Range (min … max):    29.9 ms …  67.2 ms    79 runs
...
Time (mean ± σ):      1.574 s ±  0.278 s    [User: 1.536 s, System: 0.381 s]
Range (min … max):    1.322 s …  2.309 s    10 runs

Summary
  cometary ran
   35.20 ± 5.86 times faster than cometary -m fix
```

**Comment by usrme** [comment]
Commit above already improves things significantly:

Private repository with ~2000 commits:

```diff
 $ hyperfine --warmup 10 --output inherit --ignore-failure 'cometary' 'cometary -m fix'
-Time (mean ± σ):      29.8 ms ±   3.8 ms    [User: 7.8 ms, System: 15.2 ms]
-Range (min … max):    24.8 ms …  43.5 ms    106 runs
+Time (mean ± σ):      32.5 ms ±   3.5 ms    [User: 9.3 ms, System: 16.2 ms]
+Range (min … max):    27.1 ms …  42.9 ms    93 runs
...
-Time (mean ± σ):      72.4 ms ±  16.1 ms    [User: 37.9 ms, System: 37.6 ms]
-Range (min … max):    51.2 ms … 102.6 ms    49 runs
+Time (mean ± σ):      32.2 ms ±   3.5 ms    [User: 9.2 ms, System: 16.2 ms]
+Range (min … max):    27.3 ms …  43.1 ms    92 runs
 
Summary
-  cometary ran
-    2.51 ± 0.58 times faster than cometary -m fix
+  cometary -m fix ran
+    1.01 ± 0.15 times faster than cometary
```

['flyctl'](https://github.com/superfly/flyctl) repository with ~6000 commits:

```diff
$ hyperfine --warmup 10 --output inherit --ignore-failure 'cometary' 'cometary -m fix'
-Time (mean ± σ):      29.9 ms ±   4.4 ms    [User: 8.1 ms, System: 14.7 ms]
-Range (min … max):    24.3 ms …  48.6 ms    91 runs
+Time (mean ± σ):      32.2 ms ±   3.5 ms    [User: 9.5 ms, System: 15.5 ms]
+Range (min … max):    28.1 ms …  41.5 ms    94 runs
...
-Time (mean ± σ):     111.4 ms ±  15.9 ms    [User: 87.4 ms, System: 36.6 ms]
-Range (min … max):    83.2 ms … 133.7 ms    26 runs
+Time (mean ± σ):      32.1 ms ±   3.5 ms    [User: 9.6 ms, System: 15.6 ms]
+Range (min … max):    27.7 ms …  43.7 ms    83 runs

 Summary
-  cometary ran
-    3.72 ± 0.77 times faster than cometary -m fix
+  cometary -m fix ran
+    1.00 ± 0.15 times faster than cometary
```

['kubernetes'](https://github.com/kubernetes/kubernetes) repository with ~117K commits:

```diff
$ hyperfine --warmup 10 --output inherit --ignore-failure 'cometary' 'cometary -m fix'
-Time (mean ± σ):      39.7 ms ±   6.8 ms    [User: 14.9 ms, System: 23.7 ms]
-Range (min … max):    29.9 ms …  67.2 ms    79 runs
+Time (mean ± σ):      48.9 ms ±   8.2 ms    [User: 18.9 ms, System: 28.0 ms]
+Range (min … max):    38.0 ms …  82.5 ms    62 runs
...
-Time (mean ± σ):      1.574 s ±  0.278 s    [User: 1.536 s, System: 0.381 s]
-Range (min … max):    1.322 s …  2.309 s    10 runs
+Time (mean ± σ):      47.1 ms ±   4.2 ms    [User: 18.9 ms, System: 27.0 ms]
+Range (min … max):    39.1 ms …  59.2 ms    59 runs

Summary
-  cometary ran
-   35.20 ± 5.86 times faster than cometary -m fix
+  cometary -m fix ran
+    1.04 ± 0.20 times faster than cometary
```

I think why `hyperfine` is reporting faster runs for the one doing the actual searching is because of measurement deviations.

But I think I can do another slight improvement by taking the changes from #8 out of the critical chain as well using the same methodology.

**Comment by usrme** [comment]
I had another patch geared up for tomorrow, but at the last moment yet another optimization caught my eye: before the user interface is presented, a [check is made](https://github.com/usrme/cometary/blob/1837f941d68ec32924a241851c5b5ff2ead64751/main.go#L27) for at least some files to be staged through a [Git command](https://github.com/usrme/cometary/blob/1837f941d68ec32924a241851c5b5ff2ead64751/git.go#L21) and if there is something staged then this information isn't passed on to the program, but instead is [calculated again](https://github.com/usrme/cometary/blob/1837f941d68ec32924a241851c5b5ff2ead64751/main.go#L37) using the [same Git command](https://github.com/usrme/cometary/blob/1837f941d68ec32924a241851c5b5ff2ead64751/git.go#L11).

I probably just wanted to go fast at that point and didn't really consider possible solutions to the inefficiency long enough, but now that I'm already doing performance improvements, I might as well tackle this as well.

**Comment by usrme** [comment]
Slight improvements with latest local iteration.

---

Private repository with ~2000 commits compared with previous iteration:

```diff
$ hyperfine --warmup 10 --output inherit --ignore-failure 'cometary' 'cometary -m fix'
-Time (mean ± σ):      32.5 ms ±   3.5 ms    [User: 9.3 ms, System: 16.2 ms]
-Range (min … max):    27.1 ms …  42.9 ms    93 runs
+Time (mean ± σ):      26.8 ms ±   7.6 ms    [User: 7.4 ms, System: 11.7 ms]
+Range (min … max):    13.8 ms …  49.3 ms    97 runs
...
-Time (mean ± σ):      32.2 ms ±   3.5 ms    [User: 9.2 ms, System: 16.2 ms]
-Range (min … max):    27.3 ms …  43.1 ms    92 runs
+Time (mean ± σ):      23.6 ms ±   7.3 ms    [User: 6.9 ms, System: 11.0 ms]
+Range (min … max):    14.2 ms …  58.2 ms    166 runs
 
Summary
   cometary -m fix ran
-    1.01 ± 0.15 times faster than cometary
+    1.14 ± 0.48 times faster than cometary
```

Private repository with ~2000 commits compared with original:

```diff
$ hyperfine --warmup 10 --output inherit --ignore-failure 'cometary' 'cometary -m fix'
-Time (mean ± σ):      29.8 ms ±   3.8 ms    [User: 7.8 ms, System: 15.2 ms]
-Range (min … max):    24.8 ms …  43.5 ms    106 runs
+Time (mean ± σ):      26.8 ms ±   7.6 ms    [User: 7.4 ms, System: 11.7 ms]
+Range (min … max):    13.8 ms …  49.3 ms    97 runs
...
-Time (mean ± σ):      72.4 ms ±  16.1 ms    [User: 37.9 ms, System: 37.6 ms]
-Range (min … max):    51.2 ms … 102.6 ms    49 runs
+Time (mean ± σ):      23.6 ms ±   7.3 ms    [User: 6.9 ms, System: 11.0 ms]
+Range (min … max):    14.2 ms …  58.2 ms    166 runs
 
Summary
-  cometary ran
-    2.51 ± 0.58 times faster than cometary -m fix
+  cometary -m fix ran
+    1.14 ± 0.48 times faster than cometary
```

---

['flyctl'](https://github.com/superfly/flyctl) repository with ~6000 commits compared with previous iteration:

```diff
$ hyperfine --warmup 10 --output inherit --ignore-failure 'cometary' 'cometary -m fix'
-Time (mean ± σ):      32.2 ms ±   3.5 ms    [User: 9.5 ms, System: 15.5 ms]
-Range (min … max):    28.1 ms …  41.5 ms    94 runs
+Time (mean ± σ):      26.5 ms ±   7.6 ms    [User: 7.5 ms, System: 11.7 ms]
+Range (min … max):    13.9 ms …  46.5 ms    96 runs
...
-Time (mean ± σ):      32.1 ms ±   3.5 ms    [User: 9.6 ms, System: 15.6 ms]
-Range (min … max):    27.7 ms …  43.7 ms    83 runs 
+Time (mean ± σ):      25.9 ms ±   8.0 ms    [User: 7.1 ms, System: 11.4 ms]
+Range (min … max):    14.6 ms …  57.8 ms    85 runs

Summary
   cometary -m fix ran
-    1.00 ± 0.15 times faster than cometary
+    1.02 ± 0.43 times faster than cometary
```

['flyctl'](https://github.com/superfly/flyctl) repository with ~6000 commits compared with original:

```diff
$ hyperfine --warmup 10 --output inherit --ignore-failure 'cometary' 'cometary -m fix'
-Time (mean ± σ):      29.9 ms ±   4.4 ms    [User: 8.1 ms, System: 14.7 ms]
-Range (min … max):    24.3 ms …  48.6 ms    91 runs
+Time (mean ± σ):      26.5 ms ±   7.6 ms    [User: 7.5 ms, System: 11.7 ms]
+Range (min … max):    13.9 ms …  46.5 ms    96 runs
...
-Time (mean ± σ):     111.4 ms ±  15.9 ms    [User: 87.4 ms, System: 36.6 ms]
-Range (min … max):    83.2 ms … 133.7 ms    26 runs
+Time (mean ± σ):      25.9 ms ±   8.0 ms    [User: 7.1 ms, System: 11.4 ms]
+Range (min … max):    14.6 ms …  57.8 ms    85 runs
 
Summary
-  cometary ran
-    3.72 ± 0.77 times faster than cometary -m fix
+  cometary -m fix ran
+    1.02 ± 0.43 times faster than cometary
```

---

['kubernetes'](https://github.com/kubernetes/kubernetes) repository with ~117K commits compared with previous iteration:

```diff
$ hyperfine --warmup 10 --output inherit --ignore-failure 'cometary' 'cometary -m fix'
-Time (mean ± σ):      48.9 ms ±   8.2 ms    [User: 18.9 ms, System: 28.0 ms]
-Range (min … max):    38.0 ms …  82.5 ms    62 runs
+Time (mean ± σ):      32.0 ms ±   3.7 ms    [User: 11.7 ms, System: 15.7 ms]
+Range (min … max):    26.4 ms …  44.1 ms    92 runs
...
-Time (mean ± σ):      47.1 ms ±   4.2 ms    [User: 18.9 ms, System: 27.0 ms]
-Range (min … max):    39.1 ms …  59.2 ms    59 runs
+Time (mean ± σ):      30.9 ms ±   3.8 ms    [User: 10.5 ms, System: 15.8 ms]
+Range (min … max):    25.9 ms …  51.1 ms    91 runs
  
Summary
   cometary -m fix ran
-    1.04 ± 0.20 times faster than cometary
+    1.04 ± 0.18 times faster than cometary
```

['kubernetes'](https://github.com/kubernetes/kubernetes) repository with ~117K commits compared with original:

```diff
$ hyperfine --warmup 10 --output inherit --ignore-failure 'cometary' 'cometary -m fix'
-Time (mean ± σ):      39.7 ms ±   6.8 ms    [User: 14.9 ms, System: 23.7 ms]
-Range (min … max):    29.9 ms …  67.2 ms    79 runs
+Time (mean ± σ):      32.0 ms ±   3.7 ms    [User: 11.7 ms, System: 15.7 ms]
+Range (min … max):    26.4 ms …  44.1 ms    92 runs
...
-Time (mean ± σ):      1.574 s ±  0.278 s    [User: 1.536 s, System: 0.381 s]
-Range (min … max):    1.322 s …  2.309 s    10 runs
+Time (mean ± σ):      30.9 ms ±   3.8 ms    [User: 10.5 ms, System: 15.8 ms]
+Range (min … max):    25.9 ms …  51.1 ms    91 runs

 Summary
-  cometary ran
-   35.20 ± 5.86 times faster than cometary -m fix
+  cometary -m fix ran
+    1.04 ± 0.18 times faster than cometary
```

**Comment by usrme** [comment]
Resolved in [release v1.6.0](https://github.com/usrme/cometary/releases/tag/v1.6.0).
